pkg/addonmanager/controllers/agentdeploy/controller.go

```diff
@@ -0,0 +1,230 @@
+}
+
+func NewAddonDeployController(
+	workClient workv1client.Interface,
+	addonClient addonv1alpha1client.Interface,
+	clusterInformers clusterinformers.ManagedClusterInformer,
+	addonInformers addoninformerv1alpha1.ManagedClusterAddOnInformer,
+	workInformers workinformers.ManifestWorkInformer,
+	agentAddons map[string]agent.AgentAddon,
+	recorder events.Recorder,
+) factory.Controller {
+	c := &addonDeployController{
+		workClient:                workClient,
+		addonClient:               addonClient,
+		managedClusterLister:      clusterInformers.Lister(),
+		managedClusterAddonLister: addonInformers.Lister(),
+		workLister:                workInformers.Lister(),
+		agentAddons:               agentAddons,
+		eventRecorder:             recorder.WithComponentSuffix(fmt.Sprintf("addon-deploy-controller")),
+	}
+
+	return factory.New().WithInformersQueueKeyFunc(
+		func(obj runtime.Object) string {
+			key, _ := cache.MetaNamespaceKeyFunc(obj)
+			return key
+		}, addonInformers.Informer()).
+		WithSync(c.sync).ToController(fmt.Sprintf("addon-deploy-controller"), recorder)
+}
+
+func (c *addonDeployController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	key := syncCtx.QueueKey()
+	klog.V(4).Infof("Reconciling addon deploy %q", key)
+
+	clusterName, addonName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		// ignore addon whose key is not in format: namespace/name
+		return nil
+	}
+
+	agentAddon, ok := c.agentAddons[addonName]
+	if !ok {
+		return nil
+	}
+
+	// Get ManagedCluster
+	managedCluster, err := c.managedClusterLister.Get(clusterName)
+	if errors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	managedClusterAddon, err := c.managedClusterAddonLister.ManagedClusterAddOns(clusterName).Get(addonName)
+	if errors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	owner := metav1.NewControllerRef(managedClusterAddon, addonapiv1alpha1.GroupVersion.WithKind("ManagedClusterAddon"))
+
+	managedClusterAddonCopy := managedClusterAddon.DeepCopy()
+	objects, err := agentAddon.Manifests(managedCluster, managedClusterAddon)
+	if err != nil {
+		return err
+	}
+	if len(objects) == 0 {
+		return nil
+	}
+
+	work, err := buildManifestWorkFromObject(clusterName, addonName, objects)
+	if err != nil {
+		return err
+	}
+	work.OwnerReferences = []metav1.OwnerReference{*owner}
+
+	// apply work
+	work, err = c.applyWork(ctx, work)
+	if err != nil {
+		meta.SetStatusCondition(&managedClusterAddonCopy.Status.Conditions, metav1.Condition{
+			Type:    "ManifestApplied",
+			Status:  metav1.ConditionFalse,
+			Reason:  "ManifestWorkApplyFailed",
+			Message: fmt.Sprintf("failed to apply manifestwork: %v", err),
+		})
+		c.addonClient.AddonV1alpha1().ManagedClusterAddOns(managedClusterAddonCopy.Namespace).UpdateStatus(
+			ctx, managedClusterAddonCopy, metav1.UpdateOptions{})
+		return err
+	}
+
+	// Update addon status based on work's status
```

Comment:
A condition with type `Progressing` is appended to ManagedClusterAddOn by klusterlet-addon-controller. See [here](https://github.com/open-cluster-management/klusterlet-addon-controller/blob/a836e6f94d19156e460f1433aa2df34a313b09ba/pkg/controller/managedclusteraddon/managed_cluster_addon_controller.go#L257) for more details. Should we support this condition type?

Reply:
it is a bit different, since it is not necessary that a manifestwork is created for the addon. It is possible that addon does not have any manifestwork.